pss/pss.go

```diff
@@ -144,681 +120,32 @@ type Pss struct {
-		return fmt.Errorf("Cannot unmarshal pubkey: %x", pubkeyid)
-	}
-	psp, ok := p.getPeerPub(pubkeyid, topic)
-	if !ok {
-		return fmt.Errorf("invalid topic '%s' for pubkey '%s'", topic.String(), pubkeyid)
-	}
-	return p.send(psp.address, topic, msg, true, common.FromHex(pubkeyid))
-}
-
-// Send is payload agnostic, and will accept any byte slice as payload
-// It generates an envelope for the specified recipient and topic,
-// and wraps the message payload in it.
-// TODO: Implement proper message padding
-func (p *Pss) send(to []byte, topic message.Topic, msg []byte, asymmetric bool, key []byte) error {
-	metrics.GetOrRegisterCounter("pss/send", nil).Inc(1)
-
-	if key == nil || bytes.Equal(key, []byte{}) {
-		return fmt.Errorf("Zero length key passed to pss send")
+		return nil, err
 	}
-	wrapParams := &crypto.WrapParams{
-		Sender: p.privateKey,
-	}
-	if asymmetric {
-		pk, err := p.Crypto.UnmarshalPublicKey(key)
-		if err != nil {
-			return fmt.Errorf("Cannot unmarshal pubkey: %x", key)
-		}
-		wrapParams.Receiver = pk
-	} else {
-		wrapParams.SymmetricKey = key
-	}
-	// set up outgoing message container, which does encryption and envelope wrapping
-	envelope, err := p.Crypto.Wrap(msg, wrapParams)
+	var tc chunk.Chunk
+	tc, err = trojan.Wrap(targets, m)
 	if err != nil {
-		return fmt.Errorf("failed to perform message encapsulation and encryption: %v", err)
-	}
-	log.Trace("pssmsg wrap done", "env", envelope, "mparams payload", hex.EncodeToString(msg), "to", hex.EncodeToString(to), "asym", asymmetric, "key", hex.EncodeToString(key))
-
-	// prepare for devp2p transport
-	pssMsgParams := message.Flags{
-		Symmetric: !asymmetric,
-	}
-	pssMsg := message.New(pssMsgParams)
-	pssMsg.To = to
-	pssMsg.Expire = uint32(time.Now().Add(p.msgTTL).Unix())
-	pssMsg.Payload = envelope
-	pssMsg.Topic = topic
-
-	p.enqueue(pssMsg)
-	return nil
-}
-
-// sendFunc is a helper function that tries to send a message and returns true on success.
-// It is set here for usage in production, and optionally overridden in tests.
-var sendFunc = sendMsg
-
-// tries to send a message, returns true if successful
-func sendMsg(p *Pss, sp *network.Peer, msg *message.Message) bool {
-	defer metrics.GetOrRegisterResettingTimer("pss/pp/send", nil).UpdateSince(time.Now())
-	var isPssEnabled bool
-	info := sp.Info()
-	for _, capability := range info.Caps {
-		if capability == p.capstring {
-			isPssEnabled = true
-			break
-		}
-	}
-	if !isPssEnabled {
-		log.Trace("peer doesn't have matching pss capabilities, skipping", "peer", info.Name, "caps", info.Caps, "peer", label(sp.BzzAddr.Address()))
-		return false
+		return nil, err
 	}
 
-	// get the protocol peer from the forwarding peer cache
-	pp, ok := p.getPeer(sp.BzzPeer.Peer)
-	if !ok {
-		log.Warn("peer no longer in our list, dropping message")
-		return false
-	}
-
-	err := pp.Send(context.TODO(), msg)
+	//SAVE trojanChunk to localstore, if already present do not throw error
```

Comment:
not entirely sure what you mean here, could you please explain this further?

Reply:
This was an internal comment on why it should not throw an error if the chunk exists.
I changed the message